src/main/scala/com/cognite/spark/datasource/EventsRelation.scala

```diff
@@ -184,7 +204,50 @@ class EventsRelation(config: RelationConfig)(@transient val sqlContext: SQLConte
     (postUpdate, postNewItems).parMapN((_, _) => ())
   }
 
-  override val fieldsWithPushdownFilter = Seq("type", "subtype", "assetIds", "startTime", "source")
+  override val fieldsWithPushdownFilter: Seq[String] =
+    Seq("type", "subtype", "assetIds", "startTime", "source")
+
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
+    val b = urlWithBody(filters, byIdUrl)
+    val u = urlsWithFilters(filters, listUrl)
+
+    if (b.body.nonEmpty) {
+      CdpEventsByIdsRdd(
+        sqlContext.sparkContext,
+        (e: EventItem) => {
+          if (config.collectMetrics) {
+            itemsRead.inc()
+          }
+          toRow(e, requiredColumns)
+        },
+        config,
+        partitionUrlBodies(b)
+      )
+    } else {
+      super.buildScan(requiredColumns, filters)
+    }
+  }
+
+  private def partitionUrlBodies(
+      eventUriWithBody: UriWithBody[Seq[RetrieveEvents]]): Seq[UriWithBody[Seq[RetrieveEvents]]] =
+    eventUriWithBody.body
+      .grouped(1000)
+      .map(UriWithBody(eventUriWithBody.uri, _))
+      .toSeq
+
+  def urlWithBody(filters: Array[Filter], uri: Uri): UriWithBody[Seq[RetrieveEvents]] = {
+    val elements = for {
+      field <- Seq("id", "externalId")
+      filter <- filters
+      id <- getFilter(filter, field)
+    } yield (field, id)
```

Comment:
Seems we can just yield `EventId(id.toLong)` here instead, if we skip the support for EventExternalId.